### PR TITLE
wasm: refactor EnvoyWasmVmIntegration

### LINKED ISSUE #13835: wasm: shared data and shared queue leak on VM change
`vm_id` is used as part of the key in the shared data and shared queue. When a VM is gone, the state is not cleaned up.
Ref: https://github.com/istio/istio/issues/27783

### PR DESCRIPTION
**Pull request by mathetake**
Commit Message: wasm: refactor EnvoyWasmVmIntegration
Additional Description: Also implement getLogLevel according to https://github.com/proxy-wasm/proxy-wasm-cpp-host/pull/118, and update  proxy-wasm-cpp-host  to the latest
Risk Level: low
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]

resolves #13835 by the update of proxy-wasm-cpp-host

### CODE REVIEW

File: source/extensions/common/wasm/wasm_vm.cc
Comment (PiotrSikora) [inline source/extensions/common/wasm/wasm_vm.cc:24]: You should use named enums and not their underlying integer values, since they are not guaranteed to match.

This should look like:
```
switch (ENVOY_LOGGER().level()) {
case spdlog::level::trace:
  return proxy_wasm::LogLevel::trace;
...
}
```

File: bazel/repository_locations.bzl
Comment (PiotrSikora) [inline bazel/repository_locations.bzl:847]: Nit: update to latest to avoid complains from dependency-shepards.

**Review by PiotrSikora — CHANGES_REQUESTED** [review]

File: source/extensions/common/wasm/wasm_vm.cc
Comment (mathetake) [reply source/extensions/common/wasm/wasm_vm.cc]: nice catch. thank you!

File: source/extensions/common/wasm/wasm_vm.cc
Comment (PiotrSikora) [inline source/extensions/common/wasm/wasm_vm.cc:39]: Note: This was previously logged at the `trace` level, since Envoy generally doesn't log request errors at the `error` level.

**Review by PiotrSikora — APPROVED** [review]
LGTM, sans the note.

File: source/extensions/common/wasm/wasm_vm.cc
Comment (mathetake) [reply source/extensions/common/wasm/wasm_vm.cc:39]: thanks, I was not sure why this was logged at trace but now it's clear. looks like integration()->error is only called during VM initialization so that's fine. Do you want me to add that comment to the code?

Comment (PiotrSikora) [reply source/extensions/common/wasm/wasm_vm.cc:39]: It's also called as part of `fail()`, so anytime there is crash in Proxy-Wasm module, etc.

I think logging it at the `error` level is fine, but I'll leave that decision to Envoy maintainers.

Comment (mathetake) [reply source/extensions/common/wasm/wasm_vm.cc:39]: > It's also called as part of fail()

right. let's wait for @lizan to review

Comment (lizan) [reply source/extensions/common/wasm/wasm_vm.cc:39]: In general we shouldn't log anything at higher level than debug in request path/request data unless Envoy would crash, so for `fail()`path it should be debug level.

Comment (mathetake) [reply source/extensions/common/wasm/wasm_vm.cc:39]: hmm, `fail` is called in request path only when Wasm VM crashes (trap instruction is executed) and it stop running 
https://github.com/proxy-wasm/proxy-wasm-cpp-host/blob/master/src/v8/v8.cc#L771

If I were an Wasm extension developer, then I would definitely want to have the backtrace (emitted by `fail`) and investigate what's going on since all incoming streams are immediately closed which is the urgent issue.

But I understand the point of not debug/trace logging unless Envoy crashes.

Comment (lizan) [reply source/extensions/common/wasm/wasm_vm.cc:39]: OK seems that log doesn't log any request data itself but just where it fails, so this should be good to go.

Comment (mathetake) [reply source/extensions/common/wasm/wasm_vm.cc:39]: cool!  :100:

**Pull request merged**
